### PR TITLE
[Playground][Frontend] Tags filter for Examples Catalog (#22074)

### LINKED ISSUE #22074: [Playground] Implement tag based filtering feature on the frontend
### What needs to happen?

Improve type filter (all, example, kata, unit test) to the tag based filter, that can use any custom tags.
Снимок экрана 2022-06-28 в 16 45 20

### Issue Priority

Priority: 3

### Issue Component

Component: beam-playground

### PR DESCRIPTION
**Pull request by nausharipov**
Resolves #22074 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/frontend/lib/pages/playground/states/example_selector_state.dart [hunk 1]
```diff
@@ -51,56 +85,56 @@ class ExampleSelectorState with ChangeNotifier {
...
+    final byTags = filterExamplesByTags(byType);
+    final byName = filterExamplesByName(byTags);
+    return byName;
   }
 
-  List<ExampleBase> _sortCategoryExamples(List<ExampleBase> examples) {
-    final isAllFilterType = selectedFilterType == ExampleType.all;
-    final isFilterTextEmpty = filterText.isEmpty;
-    if (isAllFilterType && isFilterTextEmpty) {
+  List<ExampleBase> filterExamplesByTags(List<ExampleBase> examples) {
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/playground/states/example_selector_state.dart:108]: `@visibleForTesting` here and below.

File: playground/frontend/lib/pages/playground/states/example_selector_state.dart [hunk 2]
```diff
@@ -51,56 +85,56 @@ class ExampleSelectorState with ChangeNotifier {
...
+    }
     return examples.where((element) => element.type == type).toList();
   }
 
-  List<ExampleBase> sortExamplesByName(
-    List<ExampleBase> examples,
-    String name,
-  ) {
+  List<ExampleBase> filterExamplesByName(List<ExampleBase> examples) {
+    if (searchText.isEmpty) {
```
Comment (alexeyinkin) [inline playground/frontend/lib/pages/playground/states/example_selector_state.dart:132]: `_searchText` here and below.

File: playground/frontend/playground_components/lib/src/cache/example_cache.dart [hunk 3]
```diff
@@ -114,6 +114,8 @@ class ExampleCache extends ChangeNotifier {
       name: result.files.first.name,
       path: id,
       description: '',
+      // TODO(nausharipov): should tags be empty?
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/cache/example_cache.dart:117]: Yes. You may delete this comment.

File: playground/frontend/playground_components/lib/src/controllers/example_loaders/content_example_loader.dart [hunk 4]
```diff
@@ -37,6 +37,8 @@ class ContentExampleLoader extends ExampleLoader {
     name: descriptor.name ?? 'Embedded_Example',
     path: '',
     description: '',
+    // TODO(nausharipov): should tags be empty?
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/lib/src/controllers/example_loaders/content_example_loader.dart:40]: Yes. You may delete this comment.

File: playground/frontend/playground_components/pubspec.lock [hunk 5]
```diff
@@ -0,0 +1,873 @@
+# Generated by pub
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/pubspec.lock:1]: Please delete this file. Libraries should not version-control `pubspec.lock`.
I will soon add it to `.gitignore`.

File: playground/frontend/playground_components/test/src/common/categories.dart [hunk 6]
```diff
@@ -18,14 +18,16 @@
 
 import 'dart:collection';
 
-import 'package:playground_components/src/models/category_with_examples.dart';
-import 'package:playground_components/src/models/sdk.dart';
+import 'package:playground_components/playground_components.dart';
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/test/src/common/categories.dart:21]: Please import specific files and not this bulk library.

File: playground/frontend/playground_components/test/src/common/categories.dart [hunk 7]
```diff
@@ -18,14 +18,16 @@
...
 
 import 'examples.dart';
 
 final categoriesMock = [
   CategoryWithExamples(title: 'Sorted', examples: [exampleMock1]),
-  CategoryWithExamples(title: 'Unsorted', examples: [exampleMock2]),
+  CategoryWithExamples(
+    title: 'Unsorted',
+    examples: [exampleMock1, exampleMock2, exampleMock2, exampleMock2],
```
Comment (alexeyinkin) [inline playground/frontend/playground_components/test/src/common/categories.dart:29]: Are they repeated to test that 'tag2' is more frequent than 'tag1'? If so, explain in a comment please.

File: playground/frontend/test/pages/playground/states/example_selector_state_test.dart [hunk 8]
```diff
@@ -118,12 +118,30 @@ void main() {
       expect(state.categories, examplesSortedByTypeMock);
       expect(exampleCache.categoryListsBySdk, exampleCache.categoryListsBySdk);
     });
-    state.sortExamplesByType(unsortedExamples, ExampleType.kata);
+    state.filterExamplesByType(unsortedExamples, ExampleType.kata);
+  });
+
+  // TODO(nausharipov): replace all 'sort' with 'filter' if ok
```
Comment (alexeyinkin) [inline playground/frontend/test/pages/playground/states/example_selector_state_test.dart:124]: Do replace please.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: playground/frontend/lib/pages/playground/states/example_selector_state.dart [hunk 9]
```diff
@@ -22,27 +22,61 @@ import 'package:playground_components/playground_components.dart';
...
+  }
+
+  List<String> _getTagsSortedByPopularity(
+    List<CategoryWithExamples> categories,
+  ) {
+    Map<String, int> tagsPopularity = {};
+    for (var category in categories) {
+      for (var example in category.examples) {
+        for (var tag in example.tags) {
+          if (tagsPopularity[tag] != null) {
```
[the remark this answers is not in the document]
Comment (nausharipov) [reply playground/frontend/lib/pages/playground/states/example_selector_state.dart:65]: Saved parentheses for readability.

File: playground/frontend/lib/modules/examples/components/filter/filter.dart [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+class _Tags extends StatelessWidget {
+  const _Tags({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    return Consumer<ExampleSelectorState>(
+      builder: (context, state, child) => Padding(
+        padding: const EdgeInsets.symmetric(vertical: kMdSpacing),
+        child: SingleChildScrollView(
```
[the remark this answers is not in the document]
Comment (nausharipov) [reply playground/frontend/lib/modules/examples/components/filter/filter.dart:86]: Screen Shot 2022-10-09 at 16 27 39

**Review by alexeyinkin — APPROVED** [review]
LGTM (internal review).

**Comment by nausharipov** [comment]
R: @damondouglas

**Review by damondouglas — APPROVED** [review]
@pabloem LGTM
Thank you @nausharipov and @alexeyinkin :-)

**Comment by pabloem** [comment]
lgtm!

**Pull request merged**